Sources/GRPCCore/Transport/InProcessClientTransport.swift

```diff
@@ -0,0 +1,153 @@
+    case closed
+  }
+
+  public typealias Inbound = RPCAsyncSequence<RPCResponsePart>
+  public typealias Outbound = RPCWriter<RPCRequestPart>.Closable
+
+  public var retryThrottle: RetryThrottle
+
+  private let executionConfigurations: ClientRPCExecutionConfigurationCollection
+  private let state: LockedValueBox<State>
+
+  public init(
+    server: InProcessServerTransport,
+    executionConfigurations: ClientRPCExecutionConfigurationCollection
+  ) {
+    self.retryThrottle = .init(maximumTokens: 10, tokenRatio: 0.1)
+    self.executionConfigurations = executionConfigurations
+    self.state = .init(.unconnected(server))
+  }
+
+  /// Establish and maintain a connection to the remote destination.
+  ///
+  /// Maintains a long-lived connection, or set of connections, to a remote destination.
+  /// Connections may be added or removed over time as required by the implementation and the
+  /// demand for streams by the client.
+  ///
+  /// Implementations of this function will typically create a long-lived task group which
+  /// maintains connections. The function exits when all open streams have been closed and new connections
+  /// are no longer required by the caller who signals this by calling ``close()``, or by cancelling the
+  /// task this function runs in.
+  ///
+  /// - Parameter lazily: This parameter is ignored in this implementation.
+  public func connect(lazily: Bool) async throws {
+    try self.state.withLockedValue { state in
+      switch state {
+      case .unconnected(let server):
+        state = .connected(server)
+        _ = server.listen()
+      case .connected:
+        throw RPCError(
+          code: .failedPrecondition,
+          message: "Already connected to server."
+        )
+      case .closed:
+        throw RPCError(
+          code: .failedPrecondition,
+          message: "Can't connect to server, transport is closed."
+        )
+      }
+    }
+  }
+
+  public func close() {
+    self.state.withLockedValue { state in
+      switch state {
+      case .unconnected(let server):
+        state = .closed
+        server.stopListening()
+      case .connected(let server):
+        state = .closed
+        server.stopListening()
+      case .closed:
+        ()
+      }
+    }
+  }
+
+  public func withStream<T>(
+    descriptor: MethodDescriptor,
+    _ closure: (RPCStream<Inbound, Outbound>) async throws -> T
+  ) async throws -> T {
+    let request = RPCAsyncSequence<RPCRequestPart>.makeBackpressuredStream(watermarks: (16, 32))
+    let response = RPCAsyncSequence<RPCResponsePart>.makeBackpressuredStream(watermarks: (16, 32))
+
+    let clientStream = RPCStream(
+      descriptor: descriptor,
+      inbound: response.stream,
+      outbound: request.writer
+    )
+
+    let serverStream = RPCStream(
+      descriptor: descriptor,
+      inbound: request.stream,
+      outbound: response.writer
+    )
+
+    let error: RPCError? = try self.state.withLockedValue { state in
```

Comment:
At the moment we have two error paths: this is throwing _and_ returns an `RPCError?`. We should only have `RPCError?` here, I think. For non `RPCError`s thrown by `acceptStream` we can return an `.unknown` status code.

Alternatively we can return a `Result<Void, RPCError>` here which is roughly the same as `RPCError?`